core/src/main/java/io/undertow/server/handlers/proxy/RouteIteratorFactory.java

```diff
@@ -0,0 +1,131 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.undertow.server.handlers.proxy;
+
+import java.nio.CharBuffer;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Factory for route/affinity iterator parser. This implementation lazily parses routes while supporting ranked routing. The
+ * iterator never creates new String instances but returns a CharSequence wrapper from the existing session ID.
+ *
+ * @author Radoslav Husar
+ */
+public class RouteIteratorFactory {
+
+    public enum ParsingCompatibility {
+        MOD_JK,
+        MOD_CLUSTER,
+    }
+
+    private final ParsingCompatibility parsing;
+    private final String delimiter;
+
+    /**
+     * @param parsingCompatibility parsing compatibility behavior
+     * @param rankedRouteDelimiter String sequence to split routes at to support ranked affinity parsing; {@code null} disables the support
+     */
+    public RouteIteratorFactory(ParsingCompatibility parsingCompatibility, String rankedRouteDelimiter) {
+        this.parsing = parsingCompatibility;
+        this.delimiter = rankedRouteDelimiter;
+    }
+
+    /**
+     * Returns an {@link Iterator<CharSequence>} of routes.
+     *
+     * @param sessionId String of sessionID from the cookie/parameter possibly including encoded/appended affinity/route information
+     * @return routes iterator; never returns {@code null}
+     */
+    public Iterator<CharSequence> iterator(String sessionId) {
+        return new RouteIterator(sessionId);
+    }
+
+    private class RouteIterator implements Iterator<CharSequence> {
+
+        private final String sessionId;
+
+        private boolean nextResolved;
+        private int nextPos;
+        private CharSequence next;
+
+        RouteIterator(String sessionId) {
+            this.sessionId = sessionId;
+
+            int index = (sessionId == null) ? -1 : sessionId.indexOf('.');
+            if (index == -1) {
+                // Case where there is no routing information at all.
+                this.nextResolved = true;
+                this.next = null;
+            } else {
+                // Case where ranked route support is not enabled
+                nextPos = index + 1;
+            }
+        }
+
+        @Override
+        public boolean hasNext() {
+            resolveNext();
+
+            return next != null;
+        }
+
+        @Override
+        public CharSequence next() {
+            resolveNext();
+
+            if (next != null) {
+                CharSequence result = next;
+                nextResolved = (delimiter == null);
```

Comment:
@rhusar can you explain why nextResolved = (delimiter==null);?
In the case where there is no delimiter and parsing equals MOD_JK, this will allow iteration to return only the first element. I can see this is deliberate, because resolveNext does not update nextPos to be equal last after next is calculated if delimiter == null, but I don't understand why. 
As a result, at ModClusterTarget won't be able to iterate through all nodes available, it will only see the first node of the list, in case there is more than one node available in that list.

Reply:
Sure, here the `delimiter == null` indicates that ranked routing is not enabled, thus the route is in mod_cluster everything after first dot (sessionID.route) and for mod_jk everything between the first and second dot (sessionID.route.domain).

Reply:
The reason we can't just enable this automatically, is that if people have been using routes like `node1.1`, `node1.2`, etc in which case if we just enable the ranked routing parsing automatically, their sessions won't be sticky. So a lot of code is dealing here with compatibility.

Reply:
@rhusar I see, thanks for the explanation, it makes sense to me.